docs/design_proposals/sync-auto-support.md

```diff
@@ -0,0 +1,150 @@
+   system may sync them, one would only be required to run `./gradlew classes`.
+2. No way to tell the system to sync non build inputs. Skaffold is
+   normally only watching `.java` files for a build, in the sync case, we want
+   it to watch `.java` files, trigger a partial build, and sync `.class` files
+   so a remote server can pick up and reload the changes.
+
+## Design
+
+#### Hack it
+To get close to the functionality we want, without modifying skaffold at all, a
+Dockerfile which depends on java build outputs could be used, like:
+```
+FROM openjdk:8
+COPY build/dependencies/ /app/dependencies
+COPY build/classes/java/main/ /app/classes
+
+CMD ["java", "-cp", "/app/classes:/app/dependencies/*", "hello.Application"]
+```
+
+with a skaffold sync block that looks like:
+```
+sync:
+  manual:
+    - src: "build/classes/java/main/**/*.class"
+      dest: "/app/classes"
+      strip: "build/classes/java/main/"
+```
+
+A user's devloop then looks like this:
+
+1. run `./gradlew classes copyDependencies`
+2. run `skaffold dev`
+3. *make changes to some java file*
+4. run `./gradlew classes`
+5. *skaffold syncs files*
+
+which is far from ideal.
+
+#### Fix it?
+
+What we might want, is an API that can accept configuration that looks like:
+
+```
+sync: auto {
+  indirect:
+  - command: "./gradlew jib classes"
+    inputs:
+      - src/main/java
+      - src/main/resources
+    files:
+      - src: "target/classes"
+        dest: "/app/classes",
+      - src: "target/resources.
+        dest: "/app/resources",
+  - command: "compile_c_stuff.sh"
+    inputs:
+    - asdf.c
+    files:
+     - src: "asdf.o"
+       dest: "/libs/asdf.o"
+  direct:
+    - src: "local/computer/extrafiles"
+      dest: "/extrafiles"
+  includeTools: true
+}
+```
+
+Lets breakdown the different parts
+
+1. `sync: auto` - just a name for the mode, can be anything
+2. `indirect` - this means, you can't sync files directly, you need a command to
+   tranform them before sync. Indirect is also a list that can handle multiple
+   indirect types.
+     1. `command` - the command to run to generate syncable files
+     2. `inputs` - the inputs to watch to trigger `command`, if left empty, this
+        should just be considered the full list of inputs that are normally
+        watched for this build minus anything in `direct` (below).
+     3. `files` - the built files to sync (and where to sync them)
+3. `direct` - same as manual sync, just copy them directly, don't wait on some
+   command to transform them
+4. `includeTools` - useful for containers that don't contain, for example, `tar`
+   and similar to debug, we can use some init container to inject the necessary
+   tooling for sync.
+
+### Open Issues/Questions
+
+Please list any open questions here in the following format:
+
+**How does a tool, like Jib, configure this**
+
+Jib will export the right data, similar to how `skaffoldFiles` currently exports
```

Comment:
This kind of answers my earlier comment https://github.com/GoogleContainerTools/skaffold/pull/2901#pullrequestreview-289605330
Can we lock down what this exported data looks like?